### PR TITLE
Exclude structure type

### PR DESCRIPTION
**Pull request by james-trayford**
implement option to exclude a structure number to complement the include option, allowing isolation of satellites 

Rebased onto https://github.com/SWIFTSIM/velociraptor-python/pull/55 , will change from draft to active status once PR55 merged.

### CODE REVIEW

File: velociraptor/autoplotter/objects.py [hunk 1]
```diff
@@ -745,20 +764,46 @@ def get_quantity_from_catalogue_with_mask(
...
             )
 
+            x = x[self.structure_mask]
+            x.name = name
+        elif self.exclude_structure_type is not None:
+            # Need to create mask
+            self.structure_mask = (
+                catalogue.structure_type.structuretype != self.exclude_structure_type
+            )
+
```
Comment (JBorrow) [inline velociraptor/autoplotter/objects.py:806]:
```suggestion
            select_mask = np._s[:]
            exclude_mask = np._s[:]

            if self.select_structure_type is not None:
                if self.exclude_structure_type == self.select_structure_type:
                    raise AutoPlotterError(
                        f"Cannot simultaneously select and exclude structure"
                        " type {self.select_structure_type}"
                    )

                select_mask = catalogue.structure_type.structuretype == self.select_structure_type
                
            if self.exclude_structure_type is not None:
                exclude_mask =  catalogue.structure_type.structuretype != self.exclude_structure_type
                
            self.structure_mask = np.logical_and.reduce(
                [
                    self.structure_mask,
                    select_mask,
                    exclude_mask,
                ]
            )

            x = x[self.structure_mask]
            x.name = name
```

**Comment by JBorrow** [comment]
Looks good apart from the selection of multiple masks

File: velociraptor/autoplotter/objects.py (hunk 1 above)
Comment (wmfw23) [reply velociraptor/autoplotter/objects.py:806]: this looks more compact but doesn't work currently as far as I can tell. I figured that you meant `s_` as this exists in numpy, and explicitly import it as you do for other python modules, but beyond that unfamiliar with this approach

Comment (JBorrow) [reply velociraptor/autoplotter/objects.py:806]: Ah I see we will need to import those functions, yeah.

`s_` is basically a shortcut for turning fancy indexing into variables, so...

```
mask = np.s_[1:3]
print(np.arange(10)[mask])
> [1 2]
```

Comment (wmfw23) [reply velociraptor/autoplotter/objects.py:806]: would be happy to test something if you have a working snippet

Comment (JBorrow) [reply velociraptor/autoplotter/objects.py:806]: Does the above not work when importing numpy?

Comment (wmfw23) [reply velociraptor/autoplotter/objects.py:806]: no, doesn't select / exclude galaxies or combine with an excplicit `structure_mask` as the original

Comment (JBorrow) [reply velociraptor/autoplotter/objects.py:806]: As in, you've tested the code and it doesn't, or it doesn't as read?

Line 776 is the line that 'includes' only galaxies of that type, 779 'excludes' galaxies, and 781 combines it with the original `structure_mask`.

Comment (wmfw23) [reply velociraptor/autoplotter/objects.py:806]: as in I've tested the code and it doesn't

Comment (JBorrow) [reply velociraptor/autoplotter/objects.py:806]: Ah I suppose the problem is that the structure mask hasn't ever been created here...

**Pull request merged**